For #391: created conditional effect

paulodamaso opened the pull request:
Closes #391: Creation of conditional effects
- Created conditional effects decorator
- Created tests

SilasReinagel commented:
@paulodamaso Is this PR still relevant?

If something new is being coded, let's close this PR and open a new one later.

paulodamaso commented:
@SilasReinagel I'll maybe need architectural changes for this one, but I'm working on it

paulodamaso commented:
@SilasReinagel Implemented correctly the card:
- Changed the image to a more adequate one
- Implemented `ConditionalEffect` , which will allow us to compose cards with conditional execution
- Implemented `EffectIfAnyHealthBelowThreshold` state, which is a conditional effect that only happens when any effect target has health below a % threshold

We have reached the architecture limit for effect composition: we can't compose effects which have more parameters than the described in `EffectData` That leaves us to a maximum of two parameters, one float and one int (`NumberOfTurns`). So we can't compose two effects if both of them use the same parameter. Note that I had to use `NumberOfTurns` parameter to pass the shield quantity to be added; if this is bad, the only other way around is to make this value fixed for this conditioinal effect.

paulodamaso commented:
@SilasReinagel I've got an idea for solving some of these problems, please do not merge yet

SilasReinagel commented:
@paulodamaso Passing arbitrary args will make this impossible to use for our Designers. 

Designers should not be required to know and memorize code in order to create their cards.

Also, the current direction of this PR doesn't seem related to the original ticket.

paulodamaso commented:
@SilasReinagel The original ticket needs a update on Effect logic, so we can pass more parameters and / or compose the effects better, so my intent is to solve update the logic and test it solvin #391

paulodamaso commented:
@SilasReinagel Cleaned the PR and made the implementation based on #467

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs:
```diff
@@ -0,0 +1,29 @@
+public abstract class ConditionalEffect : Effect
+{
+    protected Effect _effect;
+    protected Member _source;
```
This same comment is on all the PRs.

Don't store `Member` and `Target`.

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/AllEffects.cs:
```diff
@@ -35,9 +35,11 @@ public static class AllEffects
 
-        //deal .5 attack for a random enemy 3 times
+
+        //If they are still below 50% life also Shield 5.
```
Dead comment?

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/AnyTargetHealthBelowThreshold.cs:
```diff
@@ -0,0 +1,17 @@
+
+    public override bool Condition()
+    {
+        return Array.Exists(_target.Members, element => element.State[TemporalStatType.HP] < element.State.MaxHP() * _threshold );
```
It's generally better to use `System.Linq`'s `.Any()` extension method for this. 

Also, for short lambdas, it's nicer to use a single-letter variable name such as `x` or `e`.
